### ISSUE TITLE
Jdk17 GridCacheLockImpl#lock bug

### ISSUE THREAD

**Issue by ye-xiaowei** [apache/ignite#10922]
jdk: 17 ignite: 2.15
In the JDK 17 environment, when lock preemption occurs, it can get stuck.

 `GridCacheLockImpl#tryAcquire`,When the acquisition fails, it tries to join a specific queue before returning false.

```java
if (fair && !isQueued(current))
  synchronizeQueue(false, current);
```

This maintained queue on the cluster can prevent other nodes from cutting in because acquiring the lock first checks whether the head of this queue belongs to the local node.

```java
LinkedList<UUID> nodes = val.getNodes();
// Barging is prohibited in fair mode unless tryLock() is called.
if (!(bargingProhibited && hasPredecessor(nodes))) {
```

However, if multiple enqueue operations occur during one lock operation, only one dequeue operation will be performed during unlock.This can result in nodes being unable to acquire the lock. 

Unfortunately, in the JDK 17 implementation, `AbstractQueuedSynchronizer#acquire(int arg)` is called multiple times, and `isQueued` does not always return true. So even though the lock is released, it cannot be acquired. The above discussion is based on a fair lock; unfair locks do not have this queue.

 Here is the test code.
```java
@Test
public void testLock() throws InterruptedException {
  int size = 3;
  Ignite[] ignites = new Ignite[3];
  for (int i = 0; i < size; i++) {
    int finI = i;
    new Thread(
      () -> ignites[finI] = 
      Ignition.start(new IgniteConfiguration().setGridName("grid" + finI))
    ).start();
  }
  waitCluster(ignites);
  IgniteLock[] igniteLocks = new IgniteLock[size];
  for (int i = 0; i < size; i++) {
    igniteLocks[i] = ignites[i].reentrantLock("test", true, true, true);
  }

  CountDownLatch latch = new CountDownLatch(2);
  try {
    igniteLocks[0].lock();
    newThreadHoldLock(igniteLocks[1], latch);
    newThreadHoldLock(igniteLocks[2], latch);
    Thread.sleep(1_000);
  } finally {
    igniteLocks[0].unlock();
  }
  assert latch.await(10, TimeUnit.SECONDS);

  latch = new CountDownLatch(1);
  newThreadHoldLock(igniteLocks[0], latch);
  assert latch.await(10, TimeUnit.SECONDS);
}

private void waitCluster(Ignite[] ignites) throws InterruptedException {
  re:
  while (true) {
    for (Ignite ignite : ignites) {
      if (ignite == null || ignite.cluster().nodes().size() != ignites.length) {
        Thread.sleep(10);
        continue re;
      }
    }
    break;
  }
}

private void newThreadHoldLock(IgniteLock lock, CountDownLatch countDownLatch) {
  new Thread(() -> {
    try {
      lock.lock();
      countDownLatch.countDown();
    } finally {
      lock.unlock();
    }
  }).start();
}
```

**Comment by ivandasch** [comment]
@ye-xiaowei Hi! Many thanks for your report and a great reproducer. This bug will be fixed, see [here](https://issues.apache.org/jira/browse/IGNITE-20730)

**Comment by sk0x50** [comment]
The fix will be available in Apache Ignite 2.16
